### PR TITLE
5821 Enhance bundle CLI entry for different bundle workflows

### PR DESCRIPTION
**Pull request by Nic-Ma**
part of #5821 .

### Description

This PR enhanced the bundle CLI entry to support different customized bundle workflows.

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/bundle/scripts.py [hunk 1]
```diff
@@ -580,102 +581,59 @@ def get_bundle_info(
...
-        init_id=init_id_,
-        run_id=run_id_,
-        final_id=final_id_,
-        tracking=tracking_,
-        **_args,
-    )
-    workflow.initialize()
-    workflow.run()
-    workflow.finalize()
+    (workflow_name,) = _pop_args(_args, workflow=ConfigWorkflow)
```
Comment (ericspod) [inline monai/bundle/scripts.py:621]: It's really key for what's going here that the default value for `workflow_name` will be `ConfigWorkflow` is `None` is passed as the argument, it's worth commenting here that that's the case.

File: monai/bundle/scripts.py [hunk 2]
```diff
@@ -580,102 +581,59 @@ def get_bundle_info(
...
+    if isinstance(workflow_name, str):
+        workflow_class, has_built_in = optional_import("monai.bundle", name=f"{workflow_name}")  # search built-in
+        if not has_built_in:
+            workflow_class = locate(f"{workflow_name}")  # search dotted path
+        if workflow_class is None:
+            raise ValueError(f"can not locate specified workflow class: {workflow_name}.")
```
Comment (ericspod) [inline monai/bundle/scripts.py:627]:
```suggestion
            raise ValueError(f"Cannot locate specified workflow class: {workflow_name}.")
```

**Review by ericspod — CHANGES_REQUESTED** [review]
I have a few concerns with backwards compatibility with this change. I see what your intent is and that it does work correctly, but if someone has a command line `python -m monai.bundle run some_expr` they would need to have `python -m monai.bundle run --run_id some_expr`? Even besides that, because the arguments are being removed from the signature of the function the help output from Fire gives less information about what the arguments are, and what they are for, since all the important values get put into `**kwargs`.

**Comment by ericspod** [comment]
Also it looks like the premerge tests are failing for related reasons, that is the command line has changed.

File: monai/bundle/scripts.py [hunk 3]
```diff
@@ -580,102 +581,59 @@ def get_bundle_info(
 
 
 @deprecated_arg("runner_id", since="1.1", removed="1.3", new_name="run_id", msg_suffix="please use `run_id` instead.")
-def run(
```
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply monai/bundle/scripts.py:583]: Hi @ericspod ,

I see your point, adding another CLI entry may be a good idea, let me investigate it.

Thanks.

[the remark this answers is not in the document]
Comment (Nic-Ma) [reply monai/bundle/scripts.py]: Added a new entry in the latest commit.

Thanks.

File: monai/bundle/scripts.py (hunk 1 above)
Comment (Nic-Ma) [reply monai/bundle/scripts.py]: Sure, added in the latest commit.

Thanks.

File: monai/bundle/scripts.py (hunk 2 above)
Comment (Nic-Ma) [reply monai/bundle/scripts.py]: updated in the latest commit.

Thanks.

File: monai/bundle/scripts.py [hunk 4]
```diff
@@ -580,102 +581,59 @@ def get_bundle_info(
...
+    if isinstance(workflow_name, str):
+        workflow_class, has_built_in = optional_import("monai.bundle", name=f"{workflow_name}")  # search built-in
+        if not has_built_in:
+            workflow_class = locate(f"{workflow_name}")  # search dotted path
+        if workflow_class is None:
+            raise ValueError(f"can not locate specified workflow class: {workflow_name}.")
+    elif issubclass(workflow_name, BundleWorkflow):
+        workflow_class = workflow_name
+    else:
+        raise ValueError(f"`workflow` must be the bundle workflow class name, but got: {workflow_name}.")
```
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply monai/bundle/scripts.py]: updated in the latest commit.

Thanks.

**Comment by Nic-Ma** [comment]
Hi @ericspod ,

Thanks for your review and suggestions.
I totally updated the PR based on the comments and added tests, could you please help review it again?

Thanks in advance.

File: monai/bundle/scripts.py [hunk 5]
```diff
@@ -678,6 +679,48 @@ def run(
...
+    if isinstance(workflow_name, str):
+        workflow_class, has_built_in = optional_import("monai.bundle", name=f"{workflow_name}")  # search built-in
+        if not has_built_in:
+            workflow_class = locate(f"{workflow_name}")  # search dotted path
+        if workflow_class is None:
+            raise ValueError(f"cannot locate specified workflow class: {workflow_name}.")
+    elif issubclass(workflow_name, BundleWorkflow):
+        workflow_class = workflow_name
+    else:
+        raise ValueError(f"argument `workflow` must be the bundle workflow class name or type, got: {workflow_name}.")
```
Comment (ericspod) [inline monai/bundle/scripts.py:716]:
```suggestion
        raise ValueError(f"Argument `workflow` must be a bundle workflow class name or subclass of BundleWorkflow, got: {workflow_name}.")
```

File: monai/bundle/scripts.py [hunk 6]
```diff
@@ -678,6 +679,48 @@ def run(
...
+        workflow: specified bundle workflow name, should be a string or class, default to "ConfigWorkflow".
+        args_file: a JSON or YAML file to provide default values for this API.
+            so that the command line inputs can be simplified.
+        kwargs: arguments to instantiate the workflow class.
+
+    """
+
+    _args = _update_args(args=args_file, workflow=workflow, **kwargs)
+    _log_input_summary(tag="run", args=_args)
+    (workflow_name,) = _pop_args(_args, workflow=ConfigWorkflow)  # the default workflow name is "ConfigWorkflow"
```
Comment (ericspod) [inline monai/bundle/scripts.py:706]:
```suggestion
    workflow_name, = _pop_args(_args, workflow=ConfigWorkflow)  # the default workflow name is "ConfigWorkflow"
```

File: monai/bundle/scripts.py [hunk 7]
```diff
@@ -678,6 +679,48 @@ def run(
...
+    if isinstance(workflow_name, str):
+        workflow_class, has_built_in = optional_import("monai.bundle", name=f"{workflow_name}")  # search built-in
```
Comment (ericspod) [inline monai/bundle/scripts.py:708]:
```suggestion
        workflow_class, has_built_in = optional_import("monai.bundle", name=str(workflow_name))  # search built-in
```

File: monai/bundle/scripts.py [hunk 8]
```diff
@@ -678,6 +679,48 @@ def run(
...
+        if not has_built_in:
+            workflow_class = locate(f"{workflow_name}")  # search dotted path
```
Comment (ericspod) [inline monai/bundle/scripts.py:710]:
```suggestion
            workflow_class = locate(str(workflow_name))  # search dotted path
```

File: monai/bundle/scripts.py [hunk 9]
```diff
@@ -678,6 +679,48 @@ def run(
     workflow.finalize()
 
 
+def run_workflow(workflow: str | BundleWorkflow | None = None, args_file: str | None = None, **kwargs: Any) -> None:
+    """
+    Specify `bundle workflow` to run monai bundle components and workflows.
```
Comment (ericspod) [inline monai/bundle/scripts.py:684]: We should elaborate a little more here on what customised workflow classes would be and how they would operate. What's required here is to create a subclass of `BundleWorkflow` then pass it in or mention it by name, which you can on the command line. We should just state the user needs to provide that subclass and be available to import, the documentation for how to do that is elsewhere.

**Review by ericspod — APPROVED** [review]
Looks good, I have very minor style suggestions. I see you're using `pydoc.locate` here and I wonder if we should have our own function in `monai.utils.module` for doing this. This is a topic for some other time and place but I had used something like this in the past that worked well enough:

```python
def find_type_def(qualified_name: str):
    if "." not in qualified_name:
        modname, defname = "builtins", qualified_name
    else:
        modname, defname = qualified_name.rsplit(".", 1)

    mod = __import__(modname)
    return getattr(mod, defname)
```

File: monai/bundle/scripts.py (hunk 7 above)
Comment (Nic-Ma) [reply monai/bundle/scripts.py]: Updated in the latest commit, thanks.

File: monai/bundle/scripts.py (hunk 8 above)
Comment (Nic-Ma) [reply monai/bundle/scripts.py]: Updated in the latest commit, thanks.

File: monai/bundle/scripts.py (hunk 9 above)
Comment (Nic-Ma) [reply monai/bundle/scripts.py:684]: Sure, added doc-string in the latest commit, thanks.

**Comment by Nic-Ma** [comment]
Hi @ericspod ,

Thanks for your review, I updated the PR accordingly.
And for the `locate` suggestion, I created a ticket to discuss and track: https://github.com/Project-MONAI/MONAI/issues/6275.
Because we use `locate` in several places of current codebase:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/networks/nets/flexible_unet.py#L16

https://github.com/Project-MONAI/MONAI/blob/dev/monai/utils/module.py#L244

https://github.com/Project-MONAI/MONAI/blob/dev/monai/transforms/io/array.py#L208

If we want to change them together, maybe better to do a separate PR?

Thanks.

**Comment by Nic-Ma** [comment]
Hi @wyli ,

Do you know the root cause of Blossom system error here?

Thanks in advance.

**Pull request merged**